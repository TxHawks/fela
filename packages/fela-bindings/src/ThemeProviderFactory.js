/* @flow */
import { shallowEqual } from 'recompose'
import { objectEach } from 'fela-utils'

import createTheme from './createTheme'

export default function ThemeProviderFactory(
  BaseComponent: any,
  renderChildren: Function,
  statics?: Object
): any {
  class ThemeProvider extends BaseComponent {
    theme: Object

    constructor(props: Object, context: Object) {
      super(props, context)

      const previousTheme = !props.overwrite && this.context.theme
      this.theme = createTheme(props.theme, previousTheme)
    }

    componentWillReceiveProps(nextProps: Object): void {
      this.theme.update(nextProps.theme)
    }

    shouldComponentUpdate(nextProps: Object): boolean {
      return shallowEqual(this.props.theme, nextProps.theme) === false
    }

    getChildContext(): Object {
      return {
        theme: this.theme
      }
    }

    render(): Object {
      return renderChildren(this.props.children)
    }
  }

  if (statics) {
    objectEach(statics, (value, key) => {
      ThemeProvider[key] = value
    })
  }

  return ThemeProvider
}
