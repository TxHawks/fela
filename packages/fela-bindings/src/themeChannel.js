export const THEME_CHANNEL = '_FELA_THEME_'
