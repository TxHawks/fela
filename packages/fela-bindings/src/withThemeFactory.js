/* @flow */
import { hoistStatics } from 'fela-utils'

export default function withThemeFactory(
  BaseComponent: any,
  createElement: Function,
  contextTypes?: Object
): Function {
  return function withTheme(component: Object): Object {
    class WithTheme extends BaseComponent {
      state: {
        theme: Object
      }

      unsubscribe: ?Function

      constructor(props: Object, context: Object) {
        super(props, context)

        this.state = {
          theme: context.theme ? context.theme.get() : {}
        }
      }

      componentWillMount() {
        if (this.context.theme) {
          this.unsubscribe = this.context.theme.subscribe(properties =>
            this.setState({
              theme: properties
            })
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe) {
          this.unsubscribe()
        }
      }

      render() {
        const { innerRef, ...passProps } = this.props

        if (innerRef) {
          if (component._isFelaComponent) {
            passProps.innerRef = innerRef
          } else {
            passProps.ref = innerRef
          }
        }

        return createElement(component, {
          ...passProps,
          theme: this.state.theme
        })
      }
    }

    if (contextTypes) {
      WithTheme.contextTypes = contextTypes
    }

    return hoistStatics(WithTheme, component)
  }
}
