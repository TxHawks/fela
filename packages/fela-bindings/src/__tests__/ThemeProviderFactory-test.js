import React, { createElement, Component, Children } from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import withThemeFactory from '../withThemeFactory'
import ThemeProviderFactory from '../ThemeProviderFactory'
import { THEME_CHANNEL } from '../themeChannel'

const withTheme = withThemeFactory(Component, createElement, {
  [THEME_CHANNEL]: PropTypes.object,
})

const ThemeProvider = ThemeProviderFactory(
  Component,
  children => Children.only(children),
  {
    propTypes: {
      theme: PropTypes.object.isRequired,
      overwrite: PropTypes.bool,
    },
    childContextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    contextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    defaultProps: {
      overwrite: false,
    },
  }
)

describe('Using the ThemeProvider', () => {
  it('should pass the theme to rule props', () => {
    const RenderTheme = withTheme(({ theme }) => (
      <div>{JSON.stringify(theme)}</div>
    ))

    const wrapper = mount(
      <ThemeProvider theme={{ color: 'red' }}>
        <RenderTheme />
      </ThemeProvider>
    )

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should update child nodes if the theme updates', () => {
    const RenderTheme = withTheme(({ theme }) => (
      <div>{JSON.stringify(theme)}</div>
    ))

    const wrapper = mount(
      <ThemeProvider theme={{ color: 'red' }}>
        <RenderTheme />
      </ThemeProvider>
    )

    wrapper.setProps({ theme: { color: 'blue' } })

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should update child nodes through shouldComponentUpdate', () => {
    const RenderTheme = withTheme(({ theme }) => (
      <div>{JSON.stringify(theme)}</div>
    ))

    class NoUpdate extends Component {
      shouldComponentUpdate() {
        return false
      }

      render() {
        return this.props.children
      }
    }

    const wrapper = mount(
      <ThemeProvider theme={{ color: 'red' }}>
        <NoUpdate>
          <RenderTheme />
        </NoUpdate>
      </ThemeProvider>
    )

    wrapper.setProps({ theme: { color: 'blue' } })

    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
