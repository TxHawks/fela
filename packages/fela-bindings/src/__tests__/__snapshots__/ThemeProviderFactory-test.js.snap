// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using the ThemeProvider should pass the theme to rule props 1`] = `
<ThemeProvider
  overwrite={false}
  theme={
    Object {
      "color": "red",
    }
  }
>
  <WithTheme>
    <Component
      theme={
        Object {
          "color": "red",
        }
      }
    >
      <div>
        {"color":"red"}
      </div>
    </Component>
  </WithTheme>
</ThemeProvider>
`;

exports[`Using the ThemeProvider should update child nodes if the theme updates 1`] = `
<ThemeProvider
  overwrite={false}
  theme={
    Object {
      "color": "blue",
    }
  }
>
  <WithTheme>
    <Component
      theme={
        Object {
          "color": "blue",
        }
      }
    >
      <div>
        {"color":"blue"}
      </div>
    </Component>
  </WithTheme>
</ThemeProvider>
`;

exports[`Using the ThemeProvider should update child nodes through shouldComponentUpdate 1`] = `
<ThemeProvider
  overwrite={false}
  theme={
    Object {
      "color": "blue",
    }
  }
>
  <NoUpdate>
    <WithTheme>
      <Component
        theme={
          Object {
            "color": "blue",
          }
        }
      >
        <div>
          {"color":"blue"}
        </div>
      </Component>
    </WithTheme>
  </NoUpdate>
</ThemeProvider>
`;
